add "GET /states" REST API call that dumps the current component state

leo-b opened the pull request:
# What does this implement/fix?

This PR adds a `/states` REST API call that dumps the current component state as a JSON array.

As described in the feature request, this is useful for feeding sensor values into a monitoring system like Zabbix or for a quick status overview on the commandline like:

`curl -s http://devname/states | python3  -m json.tool`

Unfortunately I am not a C++ developer, so I'd appreciate any feedback.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/feature-requests/issues/1455

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#4003

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

leo-b committed:
add /states REST API call that dumps the current component state in a json array

leo-b committed:
actually return a json array

leo-b committed:
make lint checker happy

leo-b committed:
make lint checker happier

leo-b committed:
fix CI checks part 3

leo-b committed:
pass string as const reference instead of copying

nagyrobi commented:
Don't forget to update doc: https://esphome.io/web-api/index.html

OttoWinter commented:
The output of this command could be very large if users have lots of entities - and it's likely either the json serialization or web server library can't handle it well (might even crash). In the documentation please note that

leo-b commented:
> The output of this command could be very large if users have lots of entities - and it's likely either the json serialization or web server library can't handle it well (might even crash).

JSON objects are created in per component chunks, so I think serialization shouldn't be a probem. And actually I hoped that by using `AsyncResponseStream` and printing out small portions, memory wouldn't be an issue. But I guess you'll know it better than me. ;-)

Anyways I have added a note to the documentation PR.

max246 commented:
@jesserockz  you merged into web-states, is this PR still ok or what should happens?

leo-b commented:
Is there anything I can do to increase the chance of this PR being considered for merge?

nagyrobi commented:
Well I guess conflicts appeared over time need to be resolved...

leo-b committed:
add new has_connected_client() method to states.h

leo-b committed:
whitespace and clang-format cleanup

leo-b commented:
> Well I guess conflicts appeared over time need to be resolved...
> Op doesn't point to the link to the docs PR...

I have rebased the PR to current dev and added a link to the docs.
I'd appreciate if some reviewer would give this feature request a new lease of life. :-)

jesserockz commented:
Everytime I visit the "page" I get this warning in the logs

```
[13:12:12][D][esp-idf:000][httpd]: W (65505) httpd_uri: httpd_uri: uri handler execution failed

[13:12:12][D][esp-idf:000][httpd]: W (65521) httpd_uri: httpd_uri: uri handler execution failed

[13:12:12][D][esp-idf:000][httpd]: W (65540) httpd_uri: httpd_uri: uri handler execution failed
```

leo-b committed:
use application/json as mime type instead of text/json

leo-b commented:
> Everytime I visit the "page" I get this warning in the logs
> ```
> [13:12:12][D][esp-idf:000][httpd]: W (65505) httpd_uri: httpd_uri: uri handler execution failed
> ```

Unfortunately I didn't have a chance to test with the ESP-IDF framework yet. I'll get an ESP32 device and try to reproduce it with ESP-IDF. Maybe some difference in AsyncResponseStream handling?

Does the GET `/states` request actually return anything?

Which components do you use in your test config? Maybe it includes one I didn't test.

I just changed the content-type from text/json to application/json, but that won't make any difference.

leo-b commented:
Does the `/events ` call work?

jesserockz commented:
> Does the GET `/states` request actually return anything?

Yes it returns everything as expected. Was simply a warning that was not there before and is not there visiting the index page or any of the other rest endpoints.

> Which components do you use in your test config? Maybe it includes one I didn't test.

```yaml
demo:
```

It adds lots of entities. Good for testing =)
https://esphome.io/components/demo

leo-b commented:
> Yes it returns everything as expected. Was simply a warning that was not there before and is not there visiting the index page or any of the other rest endpoints.

Hmm - strange.
I just tested on an ESP32 using esp-idf and some 43 sensors (including the demo component).
I cannot reproduce the warning.
Logging is at level `VERY_VERBOSE`.

The only output I get is:

```
[21:46:07][VV][web_server_idf:106][httpd]: Enter AsyncWebServer::request_handler. method=1, uri=/states
```

leo-b committed:
remove ifdef USE_ARDUINO

leo-b commented:
> Everytime I visit the "page" I get this warning in the logs
> [13:12:12][D][esp-idf:000][httpd]: W (65505) httpd_uri: httpd_uri: uri handler execution failed

I just rebased the PR to current dev and tested it on esp-idf and adruino. I can still see no warning. :-)

TCB13 commented:
> The output of this command could be very large if users have lots of entities - and it's likely either the json serialization or web server library can't handle it well (might even crash). In the documentation please note that

This is what's called bikeshedding.

The current status of the REST API is bad. It makes no sense to force people to run 20 HTTP calls to get the status of 20 sensors and while a large dataset might crash something it would fix the problem for everyone else.

TCB13 commented:
Any updates on this? What can I do to get this merged? Thank you.
